mini-feat(docs) Improvements to the layer documentation (#2390)
zmkfirmware/zmk: docs/docs/features/keymaps.mdx

---

import KeymapExample from "../keymap-example.md";

> import KeymapExampleFile from "../keymap-example-file.md";

Reviewer: It looks like we can delete this file now.

Author: The file is still in use in the new shield guide. I thought about moving the file, or putting the content into the new shield guide. What are your thoughts?

Reviewer: Ah, good catch thanks. I'd say let's leave it for now.

## Layers

Like many mechanical keyboard firmwares, ZMK keymaps are composed of a collection of layers, with a
minimum of at least one layer that is the default, usually near the bottom of the "layer stack". Each layer
in ZMK contains a set of bindings that bind a certain behavior to a certain key position in that layer.

> |                                                   ![Diagram of three layers](../assets/features/keymaps/layer-diagram.png)                                                    |

Reviewer: This should really be replaced as well, at least with one that doesn't use the `KC_` prefix. But maybe for another PR.

Author: I am not artistically talented, so someone else will need to sort that out I think.

Reviewer: Yeah, we could ask Thomas@splitkb to share the source (he was the original author) or reproduce it otherwise.

| :---------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
| _A simplified diagram showing three layers. The layout of each layer is the same (they all contain four keys), but the behavior bindings within each layer can be different._ |

All layers are assigned and referred to by a natural number, with the base layer being layer `0`. It is common to [use the C preprocessor to "name" layers](../behaviors/layers.md#defines-to-refer-to-layers), making them more legible.

> The default layer (usually the base layer) is always enabled. Certain bound behaviors may enable/disable additional layers.

Reviewer: ```suggestion
The default layer (the base layer with index 0) is always enabled. Certain bound behaviors may enable/disable additional layers.
```
I think this can be changed in the future if any changes regarding default layers are made.

> When a key location is pressed/released, the _highest-valued currently active_ layer is used. The press/release event is sent to the behavior bound at that position in said layer, for it to perform whatever actions it wants to in reaction to the event. The behavior can choose to "consume" the event, or "pass it along" and let the next highest-valued active layer _also_ get the event (whose behavior may continue "passing it along").

Reviewer: ```suggestion
When a key location is pressed/released, the _highest-valued currently active_ layer is used. The press/release event is sent to the behavior bound at that position in said layer, for it to perform whatever actions it wants to in reaction to the event. The behavior can choose to "consume" the event, or "pass it along" and let the next highest-valued active layer _also_ get the event (whose behavior may continue "passing it along").

Note that the _activation_ order isn't relevant for determining the priority of active layers, it is determined _only_ by the definition order.
```
This remark usually helps when trying to explain to people, so I think we should insert it somewhere like this.

> 

Reviewer: I think it would be worth adding a "tip" box here, talking about multiple base layouts. That seems to be the most common scenario where people have issues.

i.e. if you have toggled layers like qwerty vs. colemak, you should always place them first in the stack (defined earlier) so that other layers that can be activated (momentarily) through them work as expected.

This might be worth a more elaborate section such as the one mentioned in #1953, but for now a short tip block would still be helpful. If we add that, then we can simply link there from here.

Author: I agree on both points, added tip block for now. When #1953 is done we can revisit.

| _A simplified diagram showing three layers. The layout of each layer is the same (they all contain four keys), but the behavior bindings within each layer can be different._ |

All layers are assigned and referred to by a natural number, with the base layer being layer `0`. It is common to [use the C preprocessor to "name" layers](../behaviors/layers.md#defines-to-refer-to-layers), making them more legible.

The default layer (the base layer with index 0) is always enabled. Certain bound behaviors may enable/disable additional layers.

When a key location is pressed/released, the _highest-valued currently active_ layer is used. The press/release event is sent to the behavior bound at that position in said layer, for it to perform whatever actions it wants to in reaction to the event. The behavior can choose to "consume" the event, or "pass it along" and let the next highest-valued active layer _also_ get the event (whose behavior may continue "passing it along").

Note that the _activation_ order isn't relevant for determining the priority of active layers, it is determined _only_ by the definition order.

:::tip

> If you wish to use multiple base layers (with a [toggle](../behaviors/layers.md#toggle-layer)), e.g. one for qwerty and another for colemak, you will want these layers to have the lowest value possible. In other words, one should be layer `0`, and the other should be layer `1`. This allows other momentary layers activated on top of them to work with both.

Reviewer: ```suggestion
If you wish to use multiple base layers (with a [toggle](../behaviors/layers.md#toggle-layer)), e.g. one for QWERTY and another for Colemak layouts, you will want these layers to have the lowest value possible. In other words, one should be layer `0`, and the other should be layer `1`. This allows other momentary layers activated on top of them to work with both.
```

### Layers

Each layer of your keymap will be nested under the keymap node. Here is an example of a layer in a 6-key macropad.

```dts
    keymap {
        compatible = "zmk,keymap";

        default_layer { // Layer 0
// ----------------------------------------------
// |     Z     |     M     |     K     |
// |     A     |     B     |     C     |
            bindings = <
                &kp Z    &kp M    &kp K
                &kp A    &kp B    &kp C
            >;
        };

>     }

Reviewer: ```suggestion
    };
```

```

Each layer should have:

1. A `bindings` property this will be a list of [behavior bindings](../behaviors/index.mdx), one for each key position for the keyboard.
1. (Optional) A `sensor-bindings` property that will be a list of behavior bindings for each sensor on the keyboard. (Currently, only encoders are supported as sensor hardware, but in the future devices like trackpoints would be supported the same way)

### Multiple Layers

Layers are numbered in the order that they appear in keymap node - the first layer is `0`, the second layer is `1`, etc.

Here is an example of a trio of layers for a simple 6-key macropad:

<KeymapExample />

:::note

> Even if layer `1` was to be activated after `2`, layer `2` would still have priority as it is higher valued. Behaviors such as [`&to`](../behaviors/layers.md#to-layer) can be used to enable one layer _and disable all other non-default layers_, though.

Reviewer: ```suggestion
Even if layer `1` was to be activated after `2`, layer `2` would still have priority as it is higher valued. Behaviors such as [To Layer (`&to`)](../behaviors/layers.md#to-layer) can be used to enable one layer _and disable all other non-default layers_, though.
```
I am trying to be aware of future Studio users, where it isn't planned to surface the binding labels to the users at all. (We probably will need a lot more changes for that!)
